eliot-service/eliot/downloader.py

```diff
@@ -0,0 +1,157 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, you can obtain one at http://mozilla.org/MPL/2.0/.
+
+"""
+Contains code for downloading sym files from one or more sources.
+"""
+
+import markus
+
+from eliot.librequests import session_with_retries
+from eliot.libsentry import get_sentry_client
+
+
+METRICS = markus.get_metrics(__name__)
+
+
+class FileNotFound(Exception):
+    """File was not found because it doesn't exist."""
+
+
+class ErrorFileNotFound(Exception):
+    """File was not found because of possible transient error."""
+
+
+class Source:
+    """Defines a source manager
+
+    This takes a source url and then uses that to retrieve SYM files from the source.
+
+    """
+
+    def __init__(self, source_url):
+        self.source_url = source_url
+
+    def get(self, debug_filename, debug_id, filename):
+        """Retrieve a source url.
+
+        :arg str debug_filename: the debug_filename
+        :arg str debug_id: the debug_id
+        :arg str filename: the symbol filename
+
+        :returns: bytes
+
+        :raises FileNotFound: if the file cannot be found
+
+        :raises ErrorFileNotFound: if the file cannot be found because of some possibly
+            transient error like a timeout or a connection error
+
+        """
+        raise NotImplementedError
+
+
+class HTTPSource(Source):
+    """Source for HTTP/HTTPS requests."""
+
+    def __init__(self, source_url):
+        self.source_url = source_url.rstrip("/") + "/"
```

Comment:
One of the things this makes difficult is if the key should go between two parts of a `source_url` and maybe the `source_url` should be more of a template. Then you could do this:

```
source_url="https://symbols.mozilla.org/%(debug_filename)s/%(debug_id)s/%(filename)s?try"
```